### ISSUE TITLE
Gstreamer backend seems to leak memory

### ISSUE THREAD

**Issue by albertz** [beetbox/audioread#62]
[Here](https://gist.github.com/albertz/4b8874a5f9aa9bce88180871b5aef8f6) is a small script to demonstrate the issue. The memory consumption constantly grows (up to 8GB).
See [here](https://github.com/librosa/librosa/issues/681) for a discussion.

**Comment by sampsyo** [comment]
Interesting! To narrow down what's going wrong, can you please do some more investigation to narrow down the leak to specific actions in the audioread library? We might have a shot at fixing this if you can point to exactly what's being leaked.

**Comment by albertz** [comment]
See the script. Actually, the only thing I use is `audio_open`, looped over a lot of FLAC files. I call it only indirectly via `librosa.load(filename, sr=None)`, which is a very straight-forward usage of `audio_open`.

**Comment by sampsyo** [comment]
I understand, but that still doesn't point to exactly where the leak is coming from. It would be awesome to have your help investigating exactly what gets leaked and when.

**Comment by albertz** [comment]
Yes, would be nice, but not sure if I have the time now (I already spent multiple hours in debugging this issue, and need to proceed with my actual work). As there as so many issues with Gstreamer anyway, I would maybe even suggest to completely remove it. My solution for now is to use PySoundFile instead of audioread. Btw., that is also what librosa is recommending.

**Comment by sampsyo** [comment]
OK! Please check back in if you ever get the chance to help.

**Comment by ssssam** [comment]
I'm experiencing this issue with Beets 1.4.6 on Fedora 28.

I tried updating to Git master of audioread, as the unreleased version 2.1.7 contains an FD leak fix (https://github.com/beetbox/audioread/commit/72ed349c12a16ab741cb02abc4de8f2e8e7fe4ee). This change causes Beets to segfault during the import. GDB shows that it's crashing inside a `read()` call in a thread.

**Comment by RyanMarcus** [comment]
Huh, that's strange -- it looks like a race. When the process is started, it seems like the reading process is delegated to a thread (i.e. QueueReaderThread). When close is called (possibly via __del__), my change closes the FDs, but potentially leaves the reader thread running. 

I haven't tested this, but it would explain why a partial read is causing the issue.
